frame/vesting/src/lib.rs

```diff
@@ -300,72 +359,127 @@ impl<T: Config> VestedTransfer for Pallet<T> {
+
 		let schedule_amount = ensure_valid_vesting_schedule::<T>(&schedule)?;
 
-		let total_amount = Self::locked_balance(to, asset)
-			.checked_add(&schedule_amount)
-			.ok_or(ArithmeticError::Overflow)?;
+		let locked = Self::locked_balance(to, asset, VestingScheduleIdSet::All)?;
+
+		let total_amount = locked.safe_add(&schedule_amount)?;
 
 		T::Currency::transfer(asset, from, to, schedule_amount)?;
 		T::Currency::set_lock(VESTING_LOCK_ID, asset, to, total_amount)?;
-		<VestingSchedules<T>>::try_append(to, asset, &schedule)
-			.map_err(|_| Error::<T>::MaxVestingSchedulesExceeded)?;
+
+		<VestingSchedules<T>>::mutate(to, asset, |schedules| {
+			schedules
+				.try_insert(vesting_schedule_id, schedule.clone())
+				.map_err(|_| Error::<T>::MaxVestingSchedulesExceeded)
+		})?;
 
 		Self::deposit_event(Event::VestingScheduleAdded {
 			from: from.clone(),
 			to: to.clone(),
 			asset,
 			schedule,
+			vesting_schedule_id,
 		});
 
 		Ok(())
 	}
 }
 
 impl<T: Config> Pallet<T> {
-	fn do_claim(who: &AccountIdOf<T>, asset: AssetIdOf<T>) -> Result<BalanceOf<T>, DispatchError> {
-		let locked = Self::locked_balance(who, asset);
+	fn do_claim(
+		who: &AccountIdOf<T>,
+		asset: AssetIdOf<T>,
+		vesting_schedule_ids: VestingScheduleIdSet<T::VestingScheduleId, T::MaxVestingSchedules>, /* Option<T::VestingScheduleId>, */
+	) -> Result<BalanceOf<T>, DispatchError> {
+		let locked = Self::locked_balance(who, asset, vesting_schedule_ids)?;
+
 		if locked.is_zero() {
 			// cleanup the storage and unlock the fund
 			<VestingSchedules<T>>::remove(who, asset);
 			T::Currency::remove_lock(VESTING_LOCK_ID, asset, who)?;
 		} else {
 			T::Currency::set_lock(VESTING_LOCK_ID, asset, who, locked)?;
 		}
+
 		Ok(locked)
 	}
 
-	/// Returns locked balance based on current block number.
-	fn locked_balance(who: &AccountIdOf<T>, asset: AssetIdOf<T>) -> BalanceOf<T> {
-		<VestingSchedules<T>>::mutate_exists(who, asset, |maybe_schedules| {
-			let total = if let Some(schedules) = maybe_schedules.as_mut() {
-				let mut total: BalanceOf<T> = Zero::zero();
-				schedules.retain(|s| {
-					let amount = s.locked_amount(
-						frame_system::Pallet::<T>::current_block_number(),
-						T::Time::now(),
-					);
-					total = total.saturating_add(amount);
-					!amount.is_zero()
-				});
-				total
-			} else {
-				Zero::zero()
-			};
-			if total.is_zero() {
-				*maybe_schedules = None;
+	/// Returns total locked balance and balance per vesting schedule, based on current block number
+	fn locked_balance(
+		who: &AccountIdOf<T>,
+		asset: AssetIdOf<T>,
+		vesting_schedule_ids: VestingScheduleIdSet<T::VestingScheduleId, T::MaxVestingSchedules>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		<VestingSchedules<T>>::try_mutate_exists(who, asset, |maybe_schedules| {
+			match maybe_schedules {
+				Some(schedules) => {
+					let ids_to_claim: Vec<_> = match vesting_schedule_ids {
+						VestingScheduleIdSet::All => schedules.keys().copied().collect(),
```

Comment:
Nit: I think a nicer pattern here would be to do iterator syntax on the ids and map to a single small function that does the part inside the for loop from line 442.

Reply:
done ✅